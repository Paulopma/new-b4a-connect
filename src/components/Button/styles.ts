import styled, { css, DefaultTheme } from 'styled-components'
import { ButtonColors, ButtonProps } from '.'
import { darken } from 'polished'

type WrapperProps = {
  hasIcon: boolean
} & Pick<ButtonProps, 'color' | 'fullWidth'> //podemos tb usar Omit<ButtonProps, 'children'>

const wrapperModifiers = {
  color: (theme: DefaultTheme, color: ButtonColors) => css`
    background: ${theme.colors[color]};

    &:hover {
      background: ${darken(0.1, theme.colors[color])};
    }
  `,
  fullWidth: () => css`
    width: 100%;
  `,
  withIcon: (theme: DefaultTheme) => css`
    display: inline-flex;
    align-items: center;
    justify-content: center;

    svg {
      width: 1.5rem;

      & + span {
        margin-left: ${theme.spacings.xxsmall};
      }
    }
  `
}

export const Wrapper = styled.button<WrapperProps>`
  ${({ theme, color, fullWidth, hasIcon }) => css`
    height: 5rem;
    display: inline-flex;
    align-items: center;
    justify-content: center;
    color: ${theme.colors.white};
    font-family: ${theme.font.family};
    font-size: ${theme.font.sizes.medium};
    padding: ${theme.spacings.xxsmall} ${theme.spacings.xxlarge};
    text-decoration: none;
    border: none;
    cursor: pointer;

    ${!!color && wrapperModifiers.color(theme, color)}
    ${!!fullWidth && wrapperModifiers.fullWidth()}
    ${!!hasIcon && wrapperModifiers.withIcon(theme)}
  `}
`
