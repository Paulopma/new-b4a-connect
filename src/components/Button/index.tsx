import { AnchorHTMLAttributes, ButtonHTMLAttributes } from 'react'
import * as S from './styles'

type ButtonTypes =
  | AnchorHTMLAttributes<HTMLAnchorElement>
  | ButtonHTMLAttributes<HTMLButtonElement>

export type ButtonColors = 'primary' | 'secondary' | 'tertiary'

export type ButtonProps = {
  children?: React.ReactNode
  color?: ButtonColors
  fullWidth?: boolean
  icon?: JSX.Element //quase equivalente ao React.ReactNode, mas mais abrangente ainda
  as?: React.ElementType
} & ButtonTypes

const Button = ({
  children,
  icon,
  color = 'primary',
  fullWidth = false,
  ...props
}: ButtonProps) => (
  <S.Wrapper fullWidth={fullWidth} color={color} hasIcon={!!icon} {...props}>
    {icon}
    {!!children && <span>{children}</span>}
  </S.Wrapper>
)
export default Button
