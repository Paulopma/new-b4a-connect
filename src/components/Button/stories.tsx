import { Story, Meta } from '@storybook/react/types-6-0'
import { AddShoppingCart } from '@styled-icons/material-outlined/AddShoppingCart'
import Button from '.'

export default {
  title: 'Button',
  component: Button,
  argTypes: {
    children: {
      type: 'string'
    },
    icon: {
      type: ''
    }
  }
} as Meta

export const Basic: Story = (args) => <Button {...args} />
Basic.args = {
  children: 'Inscreva-se'
}

export const withIcon: Story = (args) => <Button {...args} />
withIcon.args = {
  size: 'small',
  children: 'Inscreva-se',
  icon: <AddShoppingCart />
}

export const asLink: Story = (args) => <Button {...args} />
asLink.args = {
  children: 'Inscreva-se',
  as: 'a',
  href: '/link'
}
