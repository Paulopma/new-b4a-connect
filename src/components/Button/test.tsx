import { screen } from '@testing-library/react'
import Button from '.'
import React from 'react'
import { renderWithTheme } from 'utils/tests/helpers'
import { AddShoppingCart } from '@styled-icons/material-outlined/AddShoppingCart'

describe('<Button/>', () => {
  it('should render the correct size by default', () => {
    const { container } = renderWithTheme(<Button>inscreva-se</Button>)

    expect(screen.getByRole('button', { name: /inscreva-se/i })).toHaveStyle({
      height: '5rem',
      padding: '0.8rem 5.6rem',
      fontSize: '1.8rem'
    })
    expect(container.firstChild).toMatchSnapshot()
  })

  it('should render the primary color by default', () => {
    renderWithTheme(<Button>inscreva-se</Button>)

    expect(screen.getByRole('button', { name: /inscreva-se/i })).toHaveStyle({
      backgroundColor: '#FF005C'
    })
  })

  it('should render the secondary color if passed', () => {
    renderWithTheme(<Button color="secondary">inscreva-se</Button>)

    expect(screen.getByRole('button', { name: /inscreva-se/i })).toHaveStyle({
      backgroundColor: '#00AAFF'
    })
  })

  it('should render the tertiary color if passed', () => {
    renderWithTheme(<Button color="tertiary">inscreva-se</Button>)

    expect(screen.getByRole('button', { name: /inscreva-se/i })).toHaveStyle({
      backgroundColor: '#7F55AE'
    })
  })

  it('should render full width size if fullWidth is passed', () => {
    renderWithTheme(<Button fullWidth>inscreva-se</Button>)

    expect(screen.getByRole('button', { name: /inscreva-se/i })).toHaveStyle({
      width: '100%'
    })
  })

  it('should render an icon version if icon is passed', () => {
    renderWithTheme(
      <Button icon={<AddShoppingCart data-testid="icon" />}>inscreva-se</Button>
    )

    expect(screen.getByText(/inscreva-se/i)).toBeInTheDocument()
    expect(screen.getByTestId('icon')).toBeInTheDocument()
  })

  it('should render an icon version if icon is passed', () => {
    renderWithTheme(
      <Button icon={<AddShoppingCart data-testid="icon" />}>inscreva-se</Button>
    )

    expect(screen.getByText(/inscreva-se/i)).toBeInTheDocument()
    expect(screen.getByTestId('icon')).toBeInTheDocument()
  })

  it('should render Button as a link', () => {
    const { debug, container } = renderWithTheme(
      <Button as="a" href="/link">
        inscreva-se
      </Button>
    )

    debug(container)

    expect(screen.getByRole('link', { name: /inscreva-se/i })).toHaveAttribute(
      'href',
      '/link'
    )
  })
})
