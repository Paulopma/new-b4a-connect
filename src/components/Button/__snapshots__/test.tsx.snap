// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Button/> should render the correct size by default 1`] = `
.c0 {
  height: 5rem;
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  color: #FAFAFA;
  font-family: Montserrat,-apple-system,BlinkMacSystemFont,'Segoe UI',Roboto,Oxygen,Ubuntu,Cantarell,'Open Sans','Helvetica Neue',sans-serif;
  font-size: 1.8rem;
  padding: 0.8rem 5.6rem;
  -webkit-text-decoration: none;
  text-decoration: none;
  border: none;
  cursor: pointer;
  background: #FF005C;
}

.c0:hover {
  background: #cc004a;
}

<button
  class="c0"
  color="primary"
>
  <span>
    inscreva-se
  </span>
</button>
`
